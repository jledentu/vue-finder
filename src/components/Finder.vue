<script>
import TreeModel from "@/utils/tree-model";
import FinderList from "./FinderList";

/**
 * Render the tree of an item and its selected children.
 *
 * @param {Object} h          `createElement` object
 * @param {Object} context    Context component
 * @param {Object} item       Item to render
 * @return Rendering object
 */
function renderTree(h, context, item) {
  if (!item || !item.children || item.children.length === 0) {
    return null;
  }

  const expandedChild = item.children.find(child =>
    context.treeModel.isNodeExpanded(child.id)
  );

  const itemList = (
    <FinderList
      tree-model={context.treeModel}
      parent={item}
      items={item.children}
      selectable={context.selectable}
      drag-enabled={context.dragEnabled}
    />
  );

  return (
    <div class="list-container">
      {itemList}
      {expandedChild && renderTree(h, context, expandedChild)}
    </div>
  );
}

export default {
  name: "Finder",
  components: {
    FinderList
  },
  props: {
    /**
     * Data of the tree.
     */
    tree: {
      type: Object,
      required: true
    },
    /**
     * Enable the selection of items.
     */
    selectable: {
      type: Boolean,
      default: false
    },
    /**
     * Enable the drag & drop of items.
     */
    dragEnabled: {
      type: Boolean,
      default: false
    },
    /**
     * Function to filter displayed items.
     */
    filter: {
      type: Function,
      default: undefined
    }
  },
  data() {
    return {
      treeModel: {}
    };
  },
  watch: {
    filter: {
      handler(newFilter) {
        this.treeModel.filter = newFilter;
      }
    }
  },
  beforeCreate() {
    Object.defineProperty(this.$options.propsData, "tree", {
      configurable: false
    });
  },
  created() {
    this.treeModel = new TreeModel(this.tree, this.filter);
  },
  render(h) {
    return (
      <div class="tree-container">
        {this.treeModel && renderTree(h, this, this.treeModel.visibleTree)}
      </div>
    );
  }
};
</script>

<style lang="scss" scoped>
.tree-container {
  overflow-x: auto;
  height: 100%;

  .list-container {
    display: flex;
    height: 100%;
  }
}
</style>
