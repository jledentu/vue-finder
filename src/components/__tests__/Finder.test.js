import { mount } from "@vue/test-utils";
import Finder from "../Finder";

describe("Finder", () => {
  const tree = {
    id: "test1",
    children: [
      {
        id: "test11",
        label: "Test 11",
        selected: true,
        children: [
          {
            id: "test111",
            label: "Test 111"
          },
          {
            id: "test112",
            label: "Test 112"
          }
        ]
      },
      {
        id: "test12",
        label: "Test 12"
      }
    ]
  };

  it("should match snapshot", () => {
    const wrapper = mount(Finder, {
      propsData: {
        tree
      }
    });
    expect(wrapper).toMatchSnapshot();
  });

  it("should match snapshot with no children", () => {
    const wrapper = mount(Finder, {
      propsData: {
        tree: {
          id: "root"
        }
      }
    });
    expect(wrapper).toMatchSnapshot();
  });

  it("should match snapshot with a custom item component", () => {
    const wrapper = mount(Finder, {
      propsData: {
        tree,
        itemComponent: "span"
      }
    });
    expect(wrapper).toMatchSnapshot();
  });

  it("should match snapshot with expanded item and emit event", async () => {
    const wrapper = mount(Finder, {
      propsData: {
        tree
      }
    });

    wrapper
      .findAll(".item")
      .at(0)
      .trigger("focus");
    await wrapper.vm.$nextTick();
    expect(wrapper.emitted().expand).toEqual([
      [{ expanded: ["test1", "test11"] }]
    ]);
    expect(wrapper).toMatchSnapshot();
  });

  it("should match snapshot with initial filter", () => {
    const wrapper = mount(Finder, {
      propsData: {
        tree,
        filter() {
          return ({ id }) => id === "test12";
        }
      }
    });

    expect(wrapper).toMatchSnapshot();
  });

  it("should match snapshot with filter", async () => {
    const wrapper = mount(Finder, {
      propsData: {
        tree
      }
    });

    wrapper.setProps({
      filter: ({ id }) => id === "test12"
    });
    await wrapper.vm.$nextTick();

    expect(wrapper).toMatchSnapshot();
  });

  it("should match snapshot with sort", () => {
    const wrapper = mount(Finder, {
      propsData: {
        tree,
        sortBy: (item1, item2) =>
          item1.id > item2.id ? -1 : item1.id < item2.id ? 1 : 0
      }
    });

    expect(wrapper).toMatchSnapshot();
  });

  it("should match snapshot with default expanded", () => {
    const wrapper = mount(Finder, {
      propsData: {
        tree,
        defaultExpanded: "test112"
      }
    });

    expect(wrapper).toMatchSnapshot();
  });

  it("should match snapshot with an updated tree", async () => {
    const wrapper = mount(Finder, {
      propsData: {
        tree
      }
    });
    wrapper.setProps({
      tree: {
        id: "test3",
        children: [
          {
            id: "test31",
            label: "Test 31",
            children: [
              {
                id: "test311",
                label: "Test 311"
              },
              {
                id: "test312",
                label: "Test 312",
                selected: true
              }
            ]
          }
        ]
      }
    });
    await wrapper.vm.$nextTick();

    expect(wrapper).toMatchSnapshot();
  });

  it("should match snapshot with custom theme", () => {
    const wrapper = mount(Finder, {
      propsData: {
        tree,
        theme: {
          primaryColor: "#41b883",
          arrowColor: "#555",
          separatorColor: "#eee",
          separatorWidth: "3px",
          dropZoneBgColor: "rgba(112, 195, 112, 0.3)",
          draggedItemBgColor: "rgba(112, 195, 112, 0.6)"
        }
      }
    });
    expect(wrapper).toMatchSnapshot();
  });

  describe("Selection", () => {
    it("should match snapshot", () => {
      const wrapper = mount(Finder, {
        propsData: {
          tree,
          selectable: true
        }
      });

      expect(wrapper).toMatchSnapshot();
    });

    it("should emit a `select` event", async () => {
      const wrapper = mount(Finder, {
        propsData: {
          tree,
          selectable: true
        }
      });

      wrapper
        .findAll(".item > input[type=checkbox]")
        .at(1)
        .trigger("click");
      await wrapper.vm.$nextTick();

      expect(wrapper.emitted().select).toEqual([
        [{ selected: ["test11", "test12"] }]
      ]);
    });
  });

  describe("Drag & Drop", () => {
    it("should match snapshot", () => {
      const wrapper = mount(Finder, {
        propsData: {
          tree,
          dragEnabled: true
        }
      });

      expect(wrapper).toMatchSnapshot();
    });

    it("should emit a `move` event", async () => {
      const wrapper = mount(Finder, {
        propsData: {
          tree,
          dragEnabled: true
        }
      });

      wrapper
        .findAll(".item")
        .at(0)
        .trigger("dragstart", {
          dataTransfer: {
            setData() {}
          }
        });

      await wrapper.vm.$nextTick();

      wrapper
        .findAll(".item")
        .at(1)
        .trigger("drop");

      expect(wrapper.emitted().move).toEqual([
        [{ moved: "test11", to: "test12" }]
      ]);
    });
  });

  describe("API", () => {
    describe("#expand", () => {
      it("should expand the given item and emit the `expand` event", async () => {
        const wrapper = mount(Finder, {
          propsData: {
            tree,
            selectable: true
          }
        });

        wrapper.vm.expand("test112");
        await wrapper.vm.$nextTick();

        expect(wrapper).toMatchSnapshot();
        expect(wrapper.emitted().expand).toEqual([
          [
            {
              expanded: ["test1", "test11", "test112"]
            }
          ]
        ]);
      });
    });
  });
});
