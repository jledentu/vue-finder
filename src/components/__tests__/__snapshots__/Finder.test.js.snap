// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Finder Drag & Drop should match snapshot 1`] = `
<div class="tree-container">
  <div class="list-container">
    <div class="list">
      <div class="drop-zone"></div>
      <div draggable="true" class="item draggable">
        <!---->
        <div class="inner-item">Test 11</div>
        <div class="arrow"></div>
      </div>
      <div class="drop-zone"></div>
      <div draggable="true" class="item draggable">
        <!---->
        <div class="inner-item">Test 12</div>
        <!---->
      </div>
      <div class="drop-zone last"></div>
    </div>
  </div>
</div>
`;

exports[`Finder Selection should match snapshot 1`] = `
<div class="tree-container">
  <div class="list-container">
    <div class="list">
      <div draggable="false" class="item"><input type="checkbox">
        <div class="inner-item">Test 11</div>
        <div class="arrow"></div>
      </div>
      <div draggable="false" class="item"><input type="checkbox">
        <div class="inner-item">Test 12</div>
        <!---->
      </div>
    </div>
  </div>
</div>
`;

exports[`Finder should match snapshot 1`] = `
<div class="tree-container">
  <div class="list-container">
    <div class="list">
      <div draggable="false" class="item">
        <!---->
        <div class="inner-item">Test 11</div>
        <div class="arrow"></div>
      </div>
      <div draggable="false" class="item">
        <!---->
        <div class="inner-item">Test 12</div>
        <!---->
      </div>
    </div>
  </div>
</div>
`;

exports[`Finder should match snapshot with expanded item and emit event 1`] = `
<div class="tree-container">
  <div class="list-container">
    <div class="list">
      <div draggable="false" class="item expanded">
        <!---->
        <div class="inner-item">Test 11</div>
        <div class="arrow"></div>
      </div>
      <div draggable="false" class="item">
        <!---->
        <div class="inner-item">Test 12</div>
        <!---->
      </div>
    </div>
  </div>
</div>
`;

exports[`Finder should match snapshot with filter 1`] = `
<div class="tree-container">
  <div class="list-container">
    <div class="list">
      <div draggable="false" class="item">
        <!---->
        <div class="inner-item">Test 12</div>
        <!---->
      </div>
    </div>
  </div>
</div>
`;

exports[`Finder should match snapshot with initial filter 1`] = `
<div class="tree-container">
  <div class="list-container">
    <div class="list">
      <div draggable="false" class="item">
        <!---->
        <div class="inner-item">Test 11</div>
        <div class="arrow"></div>
      </div>
      <div draggable="false" class="item">
        <!---->
        <div class="inner-item">Test 12</div>
        <!---->
      </div>
    </div>
  </div>
</div>
`;

exports[`Finder should match snapshot with no children 1`] = `<div class="tree-container"></div>`;
